react-context/flow.txt

```diff
@@ -0,0 +1,136 @@
+containing arrays of all the key value pairs.
+(exp: name=bill&age=50 >> { [“name”, “bill”], [“age”, “50”] } )
+
+The App class renders a global contextValue object named AppContext, which we
+can send to any module as long as they deconstruct what they need from this.context.
+
+The App’s render method calls the this.renderPage() method. The renderPage
+method is deciding which page we will render, based on this.state.route.path.
+At first, because the path is an empty string, we return the <Home /> element
+to the index.jsx, to be rendered to the screen. This requires the use of the
+home.jsx module.
+
+In home.jsx, we check the status of the content we imported with AppContext.
+Now, the this of class Home contains context which was exported from App.
+Because there is currently no user in this.context.user, we return
+the <Redirect /> element with a prop of { to: “sign-in”}.
+
+In redirect.jsx, we are constructing a new URL object with the current
+window.location. Then, we are changing the URL’s hash property to the value of
+the prop to (“sign-in”). Once that is done, we are assigning the current
+window.location to the new URL we made, with the updated hash. This is
+effectively changing what we view on the screen to the #sign-in page.
+
+I don’t understand why we are back in app.jsx.
+Maybe we’re back here because the state of route changed?
+But now we are back in app.jsx and re-rendering the page, which is calling
+the this.renderPage method. But now, this.state.path is not equal to an
+empty string. It is now equal to “sing-up”. So, renderPage returns
+the <Auth /> element, from the auth.jsx file.
+
+In auth.jsx, we are pulling the data out of AppContext. Since the user is still
+null, we move on to generating the react elements of the page. Based on
+route.path being === “sign-in”, we are saying “Please Sign in”. If the path
+were “sign-up”,  we’d be saying “Make an account”. So, we start creating an
+element with the appropriate message for the user. Part of the process of
+making the element calls for the react element <AuthForm /> with the properties
+key, action, and onSignIn.
+
+In the auth-form.jsx we are importing the props passed from auth.jsx and giving
+the state object of the elements an empty username and an empty password. The
+render method of the class then generates the form needed, with the appropriate
+labels for signing in or creating an account. This form is being created with
+event listeners for change in the boxes and a button for submit. The created
+form gets returned to the auth.jsx file which then returns the entire sign-in
+page to renderPage method of the App class. Then the element gets inserted into
+the return of the render of App. We now have a sign in page on the screen.
+
+The user clicks Sign-up, causing the this.state.route to change, triggering the
+componentDidMount() of the Apps class to run again. The Apps class goes through
```

Comment:
Close! `render` runs again (`setState` ran because of the `'hashchange'` event and `setState` schedules a re-render. `componentDidUpdate` only runs _one time_, after the initial `render()` over the life of a component.